src/buffer/mod.rs

```diff
@@ -74,21 +99,65 @@ pub trait Float: Sealed {}
 impl Float for f32 {}
 impl Float for f64 {}
 
-pub trait Sealed {
+pub trait Sealed: Copy {
+    fn is_nonfinite(self) -> bool;
+    fn format_nonfinite(self) -> &'static str;
     unsafe fn write_to_ryu_buffer(self, result: *mut u8) -> usize;
 }
 
 impl Sealed for f32 {
     #[inline]
-    #[cfg_attr(feature = "no-panic", no_panic)]
+    fn is_nonfinite(self) -> bool {
+        const EXP_MASK: u32 = 0x7f800000;
+        let bits = unsafe { mem::transmute::<f32, u32>(self) };
+        bits & EXP_MASK == EXP_MASK
+    }
+
+    #[cold]
+    #[cfg_attr(feature = "no-panic", inline)]
+    fn format_nonfinite(self) -> &'static str {
+        const MANTISSA_MASK: u32 = 0x007fffff;
+        const SIGN_MASK: u32 = 0x80000000;
+        let bits = unsafe { mem::transmute::<f32, u32>(self) };
+        if bits & MANTISSA_MASK != 0 {
+            NAN
+        } else if bits & SIGN_MASK != 0 {
+            NEG_INFINITY
+        } else {
+            INFINITY
+        }
+    }
+
+    #[inline]
     unsafe fn write_to_ryu_buffer(self, result: *mut u8) -> usize {
         raw::format32(self, result)
     }
 }
 
 impl Sealed for f64 {
     #[inline]
-    #[cfg_attr(feature = "no-panic", no_panic)]
+    fn is_nonfinite(self) -> bool {
+        const EXP_MASK: u64 = 0x7ff0000000000000;
+        let bits = unsafe { mem::transmute::<f64, u64>(self) };
+        bits & EXP_MASK == EXP_MASK
```

Comment:
Interesting. Does this wind up being faster than [`!number.f64::is_finite()`](https://doc.rust-lang.org/std/primitive.f64.html#method.is_finite)?

Also, thanks so much for this! :-)

Reply:
Two things:

- Until rust-lang/rust#49896, that method is std-only. We support no_std and older compilers than that.

- The implementation here is the same speed on 1.34+ but substantially faster than f64::is_finite on older compilers which don't have https://github.com/rust-lang/rust/pull/57353.